### PR TITLE
Add support for AKS Extensions & Marketplace Offers

### LINKED ISSUE #3513: Support for AKS Marketplace offers
/kind feature

**Describe the solution you'd like**
Ability to install [Azure container marketplace offers](https://learn.microsoft.com/en-us/azure/aks/deploy-marketplace) via CAPZ managedclusters.

**Anything else you would like to add:**
[Go SDK is available here](https://github.com/Azure/azure-sdk-for-go/tree/sdk/resourcemanager/kubernetesconfiguration/armkubernetesconfiguration/v1.1.1/sdk/resourcemanager/kubernetesconfiguration/armkubernetesconfiguration/)
[Filed feature request on ASO here](https://github.com/Azure/azure-service-operator/issues/2934)

### PR DESCRIPTION
**Pull request by willie-yao**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds support for AKS Marketplace extensions

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3513

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Add support for AKS Extensions & Marketplace Offers
```

### CODE REVIEW

File: Tiltfile [hunk 1]
```diff
@@ -22,7 +22,7 @@ settings = {
     "capi_version": "v1.5.3",
     "cert_manager_version": "v1.13.1",
     "kubernetes_version": "v1.28.0",
-    "aks_kubernetes_version": "v1.26.3",
+    "aks_kubernetes_version": "v1.27.3",
```
Comment (Jont828) [inline Tiltfile:25]: Q: is this version bump what enables the support for marketplace extensions?

File: api/v1beta1/azuremanagedcontrolplane_webhook.go [hunk 2]
```diff
@@ -676,6 +683,45 @@ func (m *AzureManagedControlPlane) validateOIDCIssuerProfileUpdate(old *AzureMan
 	return allErrs
 }
 
+// validateMarketplaceExtensionsUpdate validates update to Marketplace extensions.
+func (m *AzureManagedControlPlane) validateMarketplaceExtensionsUpdate(old *AzureManagedControlPlane) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if m.Spec.MarketplaceExtensions != nil && old.Spec.MarketplaceExtensions != nil {
+		for i, extension := range m.Spec.MarketplaceExtensions {
+			if extension.Name != old.Spec.MarketplaceExtensions[i].Name {
```
Comment (Jont828) [inline api/v1beta1/azuremanagedcontrolplane_webhook.go:692]: This validation looks like it expects the list of `Spec.MarketplaceExtensions` to be in the same order since otherwise the name check will fail if you swap the order of two entries. Do you think it'd make sense to treat different permutations as identical?

File: api/v1beta1/azuremanagedcontrolplane_webhook_test.go [hunk 3]
```diff
@@ -1108,6 +1118,77 @@ func TestValidatingWebhook(t *testing.T) {
...
+									Name:      "test-plan",
+									Product:   "test-product",
+									Publisher: "test-publisher",
+								},
+							},
+						},
+					},
+				},
+			},
+			expectErr: false,
```
Comment (Jont828) [inline api/v1beta1/azuremanagedcontrolplane_webhook_test.go:1142]: Out of scope for this PR, but do you think it'd be better to change the type to `expectedErr` and match the error message?

File: azure/scope/managedcontrolplane.go [hunk 4]
```diff
@@ -903,3 +905,35 @@ func (s *ManagedControlPlaneScope) PrivateEndpointSpecs() []azure.ASOResourceSpe
...
+			Namespace:               s.Cluster.Namespace,
+			AutoUpgradeMinorVersion: extension.AutoUpgradeMinorVersion,
+			ConfigurationSettings:   extension.ConfigurationSettings,
+			ExtensionType:           extension.ExtensionType,
+			ReleaseTrain:            extension.ReleaseTrain,
+			Version:                 extension.Version,
+			Owner:                   azure.ManagedClusterID(s.SubscriptionID(), s.ResourceGroup(), s.ControlPlane.Name),
+			OwnerRef:                *metav1.NewControllerRef(s.ControlPlane, infrav1.GroupVersion.WithKind(infrav1.AzureManagedControlPlaneKind)),
+			Plan:                    *extension.Plan,
+			// AKSAssignedIdentityType: extension.AKSAssignedIdentityType,
```
Comment (Jont828) [inline azure/scope/managedcontrolplane.go:932]: Q: I know this is WIP but I'm curious if this is meant to be included.

**Review by Jont828 — COMMENTED** [review]
Nice work, this is looking good so far!

File: Tiltfile (hunk 1 above)
Comment (willie-yao) [reply Tiltfile:25]: Marketplace extensions are enabled by ASO 2.4.0. I had to bump the k8s version because I can't create AKS clusters in Tilt on 1.26.3. I think this might be due to the new ASO version but I'm not entirely sure.

File: api/v1beta1/azuremanagedcontrolplane_webhook.go (hunk 2 above)
Comment (willie-yao) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go]: Yup I realized the error in the logic there and redid the logic to use a map: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4360/files/07d7bac254be26620489fcdea39fe9d8325aad35..a4281810698097954f1e1332c922e0a6ef474225#diff-fba23483604b01adf9ec7f060a2e74fc7762cf8b9bb166451e3275389be96270R692

File: api/v1beta1/azuremanagedcontrolplane_webhook_test.go (hunk 3 above)
Comment (willie-yao) [reply api/v1beta1/azuremanagedcontrolplane_webhook_test.go:1142]: Yeah I think that would be ideal. There are a ton of test cases in this file but I still think it'll be worth doing

File: azure/scope/managedcontrolplane.go (hunk 4 above)
Comment (willie-yao) [reply azure/scope/managedcontrolplane.go]: Yeah I just haven't added the field in the type yet, although this is now fully implemented in the latest commit.

File: api/v1beta1/azuremanagedcontrolplane_types.go [hunk 5]
```diff
@@ -421,6 +421,52 @@ type OIDCIssuerProfile struct {
...
+	// AutoUpgradeMinorVersion is a flag to note if this extension participates in auto upgrade of minor version, or not.
+	// +kubebuilder:default=true
+	// +optional
+	AutoUpgradeMinorVersion *bool `json:"autoUpgradeMinorVersion,omitempty"`
+
+	// ConfigurationSettings are the name-value pairs for configuring this extension.
+	// +optional
+	ConfigurationSettings map[string]string `json:"configurationSettings,omitempty"`
+
+	// ExtensionType is the of the Extension, of which this resource is an instance of.
```
Comment (mboersma) [inline api/v1beta1/azuremanagedcontrolplane_types.go:445]:
```suggestion
	// ExtensionType is the type of the Extension of which this resource is an instance.
```

File: api/v1beta1/azuremanagedcontrolplane_types.go [hunk 6]
```diff
@@ -421,6 +421,52 @@ type OIDCIssuerProfile struct {
...
+	// It must be one of the Extension Types registered with Microsoft.KubernetesConfiguration by the Extension publisher.
+	ExtensionType *string `json:"extensionType"`
+
+	// Plan is the plan of the extension.
+	Plan *MarketplacePlan `json:"plan"`
+
+	// ReleaseTrain is the release train this extension participates in for auto-upgrade (e.g. Stable, Preview, etc.)
+	// only if autoUpgradeMinorVersion is ‘true’.
```
Comment (mboersma) [inline api/v1beta1/azuremanagedcontrolplane_types.go:453]:
```suggestion
	// This is only used if autoUpgradeMinorVersion is ‘true’.
```

File: api/v1beta1/azuremanagedcontrolplane_webhook.go [hunk 7]
```diff
@@ -687,6 +694,89 @@ func (m *AzureManagedControlPlane) validateOIDCIssuerProfileUpdate(old *AzureMan
...
+		if (extension.Version != nil && oldExtension.Version != nil) && *extension.Version != *oldExtension.Version {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("Spec", "MarketplaceExtensions", fmt.Sprintf("[%d]", i), "Version"),
+					extension.Version,
+					"field is immutable",
+				),
+			)
+		}
+		if extension.Identity != oldExtension.Identity {
```
Comment (mboersma) [inline api/v1beta1/azuremanagedcontrolplane_webhook.go:766]: This is a lot of repeated code here, but I can't think of a way to do this validation otherwise without using reflection.

File: api/v1beta1/types.go [hunk 8]
```diff
@@ -1090,3 +1090,72 @@ const (
 	// UniformOrchestrationMode treats VMs as identical instances accessible by the VMSS VM API.
 	UniformOrchestrationMode OrchestrationModeType = "Uniform"
 )
+
+// MarketplacePlan represents the plan for an AKS marketplace extension.
+type MarketplacePlan struct {
+	// Name is the user defined name of the 3rd Party Artifact that is being procured.
```
Comment (mboersma) [inline api/v1beta1/types.go:1096]:
```suggestion
	// Name is the user-defined name of the 3rd Party Artifact that is being procured.
```

File: api/v1beta1/types.go [hunk 9]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// Product is the name of the 3rd Party artifact that is being procured.
+	Product string `json:"product"`
+
+	// PromotionCode is a publisher provided promotion code as provisioned in Data Market for the said product/artifact.
```
Comment (mboersma) [inline api/v1beta1/types.go:1103]:
```suggestion
	// PromotionCode is a publisher-provided promotion code as provisioned in Data Market for the said product/artifact.
```

File: api/v1beta1/types.go [hunk 10]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+	// +optional
+	PromotionCode string `json:"promotionCode,omitempty"`
+
+	// Publisher is the name of the publisher of the 3rd Party Artifact that is being bought
```
Comment (mboersma) [inline api/v1beta1/types.go:1107]:
```suggestion
	// Publisher is the name of the publisher of the 3rd Party Artifact that is being bought.
```

File: api/v1beta1/types.go [hunk 11]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+	Publisher string `json:"publisher"`
+
+	// Version is the version of the plan.
+	// +optional
+	Version string `json:"version,omitempty"`
+}
+
+// ExtensionScope defines the scope of the aks marketplace extension, if configured.
```
Comment (mboersma) [inline api/v1beta1/types.go:1115]:
```suggestion
// ExtensionScope defines the scope of the AKS marketplace extension, if configured.
```

File: api/v1beta1/types.go [hunk 12]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+type ExtensionScope struct {
+	// ScopeType is the scope of the extension. It can be either Cluster or Namespace; but not both.
```
Comment (mboersma) [inline api/v1beta1/types.go:1117]:
```suggestion
	// ScopeType is the scope of the extension. It can be either Cluster or Namespace, but not both.
```

File: api/v1beta1/types.go [hunk 13]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+	ScopeType ExtensionScopeType `json:"scopeType"`
+
+	// ReleaseNamespace is the namespace where the extension Release must be placed, for a Cluster scoped extension.
```
Comment (mboersma) [inline api/v1beta1/types.go:1120]:
```suggestion
	// ReleaseNamespace is the namespace where the extension Release must be placed, for a Cluster-scoped extension.
```

File: api/v1beta1/types.go [hunk 14]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+	// Required for Cluster scoped extensions.
```
Comment (mboersma) [inline api/v1beta1/types.go:1121]:
```suggestion
	// Required for Cluster-scoped extensions.
```

File: api/v1beta1/types.go [hunk 15]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+	// +optional
+	ReleaseNamespace string `json:"releaseNamespace,omitempty"`
+
+	// TargetNamespace is the namespace where the extension will be created for an Namespace scoped extension.
```
Comment (mboersma) [inline api/v1beta1/types.go:1125]:
```suggestion
	// TargetNamespace is the namespace where the extension will be created for a Namespace-scoped extension.
```

File: api/v1beta1/types.go [hunk 16]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+	// Required for Namespace scoped extensions.
```
Comment (mboersma) [inline api/v1beta1/types.go:1126]:
```suggestion
	// Required for Namespace-scoped extensions.
```

File: api/v1beta1/types.go [hunk 17]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+	// +optional
+	TargetNamespace string `json:"targetNamespace,omitempty"`
+}
+
+// ExtensionScopeType defines the scope type of the aks marketplace extension, if configured.
```
Comment (mboersma) [inline api/v1beta1/types.go:1131]:
```suggestion
// ExtensionScopeType defines the scope type of the AKS marketplace extension, if configured.
```

File: api/v1beta1/types.go [hunk 18]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+type ExtensionScopeType string
+
+const (
+	// ExtensionScopeCluster ...
+	ExtensionScopeCluster ExtensionScopeType = "Cluster"
+	// ExtensionScopeNamespace ...
+	ExtensionScopeNamespace ExtensionScopeType = "Namespace"
+)
+
+// ExtensionIdentity defines the identity of the aks marketplace extension, if configured.
```
Comment (mboersma) [inline api/v1beta1/types.go:1142]:
```suggestion
// ExtensionIdentity defines the identity of the AKS marketplace extension, if configured.
```

File: api/v1beta1/types.go [hunk 19]
```diff
@@ -1090,3 +1090,72 @@ const (
...
+// +kubebuilder:validation:Enum=SystemAssigned
+type ExtensionIdentity string
+
+const (
+	// ExtensionIdentitySystemAssigned ...
+	ExtensionIdentitySystemAssigned ExtensionIdentity = "SystemAssigned"
+)
+
+// AKSAssignedIdentity defines the AKS assigned identity of the aks marketplace extension, if configured.
```
Comment (mboersma) [inline api/v1beta1/types.go:1151]:
```suggestion
// AKSAssignedIdentity defines the AKS-assigned identity of the AKS marketplace extension, if configured.
```

File: api/v1beta1/azuremanagedcontrolplane_webhook.go (hunk 7 above)
Comment (willie-yao) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:766]: What is reflection?

File: azure/services/aksextensions/spec.go [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+		Version:   ptr.To(s.Plan.Version),
+	}
+	if s.ExtensionIdentity != "" {
+		aksExtension.Spec.Identity = &asokubernetesconfigurationv1.Identity{
+			Type: (*asokubernetesconfigurationv1.Identity_Type)(ptr.To(s.ExtensionIdentity)),
+		}
+	}
+	if s.AKSAssignedIdentityType != "" {
+		aksExtension.Spec.AksAssignedIdentity = &asokubernetesconfigurationv1.Extension_Properties_AksAssignedIdentity_Spec{
+			Type: (*asokubernetesconfigurationv1.Extension_Properties_AksAssignedIdentity_Type_Spec)(ptr.To(s.AKSAssignedIdentityType)),
```
Comment (willie-yao) [inline azure/services/aksextensions/spec.go:86]: For some reason, the AKSAssignedIdentity type field is not being reflected on Azure. I think it may be overwritten by AKS, but I will need to verify the behavior. In the meantime, I have disabled the check for this field being set in the E2E test: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4360/commits/38eb084ce82692919d0f8249332b70d4e2c5e7f9

**Comment by willie-yao** [comment]
Turns out the issue with the E2E test is that the `Microsoft.KubernetesConfiguration RP` is not registered in the CNCF subscription. Retesting now.

File: Tiltfile [hunk 21]
```diff
@@ -22,8 +22,8 @@ settings = {
     "kind_cluster_name": "capz",
     "capi_version": "v1.6.0",
     "cert_manager_version": "v1.13.2",
-    "kubernetes_version": "v1.28.3",
-    "aks_kubernetes_version": "v1.28.3",
+    "kubernetes_version": "v1.28.0",
```
Comment (CecileRobertMichon) [inline Tiltfile:25]: why this change?

Comment (jackfrancis) [reply Tiltfile:25]: +1

(I assume this is cruft/rebase errata)

File: azure/services/aksextensions/aksextensions.go [hunk 22]
```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2023 The Kubernetes Authors.
```
Comment (CecileRobertMichon) [inline azure/services/aksextensions/aksextensions.go:2]: nit: it's 2024 now :)

File: api/v1beta1/azuremanagedcontrolplane_types.go [hunk 23]
```diff
@@ -421,6 +421,52 @@ type OIDCIssuerProfile struct {
 	Enabled *bool `json:"enabled,omitempty"`
 }
 
+// MarketplaceExtension represents the configuration for a marketplace extension.
+// See also [AKS doc].
```
Comment (jackfrancis) [inline api/v1beta1/azuremanagedcontrolplane_types.go:425]: Nit: is this markdown-like anchor thing (`[AKS doc]`) a go comment idiom?

File: Tiltfile (hunk 21 above)
Comment (willie-yao) [reply Tiltfile:25]: Yup, leftover from testing. I was seeing if reverting the version would fix the test like it did for fleet (it did not)

File: api/v1beta1/azuremanagedcontrolplane_types.go (hunk 23 above)
Comment (willie-yao) [reply api/v1beta1/azuremanagedcontrolplane_types.go:425]: Yes. Jon actually just explained this [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4316#discussion_r1456552607).

Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_types.go:425]: TIL!

File: azure/services/aksextensions/spec.go (hunk 20 above)
Comment (jackfrancis) [reply azure/services/aksextensions/spec.go:86]: ping @nojnhuh in case this is an ASO AKS resource edge case

**Comment by jackfrancis** [comment]
let's address the nits (copyright and version regression) otherwise lgtm!

**Comment by willie-yao** [comment]
Forgot to add webhook validation to the amcp template for extensions. Will push a commit to address this. It will re-use the same validation functions, just with a bit of re-formatting. Sorry for not doing this earlier!

File: azure/services/aksextensions/spec.go (hunk 20 above)
Comment (nojnhuh) [reply azure/services/aksextensions/spec.go:86]: I'm not sure what might be going on here besides a general Azure API quirk. @willie-yao  What exactly are you seeing? And how about in the status of the ASO resource?

Comment (willie-yao) [reply azure/services/aksextensions/spec.go:86]: When I describe the extension resource, `AKSAssignedIdentity.type` is null, but the principal ID and tenant ID are populated. I think AKS populates those fields but the user specifies the type as seen [here](https://azure.github.io/azure-service-operator/reference/kubernetesconfiguration/v1api20230501/#kubernetesconfiguration.azure.com/v1api20230501.Extension_Properties_AksAssignedIdentity_Type_Spec)

Comment (willie-yao) [reply azure/services/aksextensions/spec.go:86]: This is what I get with `AKSAssignedIdentity` set to `SystemAssigned` when I run `az k8s-extension show`
```
"aksAssignedIdentity": {
    "principalId": "9899e674-c34c-4bd1-8edf-845e5f5d3f22",
    "tenantId": null,
    "type": null
  },
```

File: controllers/azuremanagedcontrolplane_reconciler.go [hunk 24]
```diff
@@ -59,6 +60,7 @@ func newAzureManagedControlPlaneReconciler(scope *scope.ManagedControlPlaneScope
 			managedclusters.New(scope),
 			privateendpoints.New(scope),
 			resourceHealthSvc,
+			aksextensions.New(scope),
```
Comment (nojnhuh) [inline controllers/azuremanagedcontrolplane_reconciler.go:63]: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4316#discussion_r1424791178

File: azure/scope/managedcontrolplane.go [hunk 25]
```diff
@@ -912,3 +914,35 @@ func (s *ManagedControlPlaneScope) PrivateEndpointSpecs() []azure.ASOResourceSpe
...
+// AKSExtensionSpecs returns the AKS extension specs.
+func (s *ManagedControlPlaneScope) AKSExtensionSpecs() []azure.ASOResourceSpecGetter[*asokubernetesconfigurationv1.Extension] {
+	if s.AKSExtension() == nil {
+		return nil
+	}
+	extensionSpecs := make([]azure.ASOResourceSpecGetter[*asokubernetesconfigurationv1.Extension], 0, len(s.ControlPlane.Spec.MarketplaceExtensions))
+	for _, extension := range s.AKSExtension() {
+		extensionSpec := &aksextensions.AKSExtensionSpec{
+			Name:                    extension.Name,
+			Namespace:               s.Cluster.Namespace,
```
Comment (jackfrancis) [inline azure/scope/managedcontrolplane.go:932]: Sorry if this was asked earlier, why are we not allowing the user to configure the extension namespace?

Comment (willie-yao) [reply azure/scope/managedcontrolplane.go:932]: That's a mistake on my end. I think I erroneously assumed the extension has to be in the same namespace as the cluster. I'll add it as a field if it's fine to be in a different namespace.

Comment (nojnhuh) [reply azure/scope/managedcontrolplane.go:932]: All of the other ASO resources are assumed to be in the cluster namespace currently and that would be a hard requirement for #4340, so I'd vote to keep this assumption for now.

**Comment by willie-yao** [comment]
Running into the same issue as AKS Fleet with the E2E test not passing on k8s `v1.28.3`, so I'm also moving this test spec in front of the upgrade test.

Comment on the AKS Fleet PR: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4316#discussion_r1435336345

**Comment by nojnhuh** [comment]
@Jont828 PTAL

**Comment by willie-yao** [comment]
Just renamed any mentions of "Marketplace" to just "AKS" because you can deploy Azure Extensions as well as those from Marketplace. i.e. these extensions are not exclusive to just Marketplace extensions.

https://learn.microsoft.com/en-us/azure/aks/deploy-extensions-az-cli#create-extension-instance

**Comment by willie-yao** [comment]
Looks like `pull-cluster-api-provider-azure-windows-custom-builds` and `pull-cluster-api-provider-azure-windows-with-ci-artifacts` are really flaky under this PR. Not sure what could be causing it but I'll look into it. There are flakes on other PRs but they aren't as common.
